appinventor/components/src/com/google/appinventor/components/runtime/util/JavaReplaceWithMappings.java

```diff
@@ -0,0 +1,189 @@
+// -*- mode: java; c-basic-offset: 2; -*-
+// Copyright 2020  MIT, All rights reserve
+// Released under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package com.google.appinventor.components.runtime.util;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Java implementation of replace-with-mappings. Used for ease
+ * of reasoning about the solution.
+ * See runtime.scm
+ * <p>
+ * TODO: Might be better to re-implement this in Scheme/Kawa in the future.
+ * TODO: Add unit tests
+ */
+public final class JavaReplaceWithMappings {
+  /**
+   * Replaces the specified text string with the specified mappings,
+   * which is a map containing Key Value pairs where the key is the
+   * substring to replace, and the value is the value to replace the substring with.
+   * A mode parameter is also specified to indicate the order of mapping application:
+   * 0 - longest string first
+   * 1 - dictionary order
+   * 2 - earliest string first
+   *
+   * @param text     Text to apply mappings to
+   * @param mappings Map containing mappings
+   * @param mode     Mode to use for replacing mappings
+   * @return Text with the mappings applied
+   */
+  public static String replaceWithMappings(String text, Map<Object, Object> mappings, int mode) {
+    // Iterate over all the mappings
+    Iterator<Map.Entry<Object, Object>> it = mappings.entrySet().iterator();
+
+    // Construct a new map for <String, String> mappings in order to support
+    // look-ups for non-pure String values (e.g. numbers)
+    Map<String, String> stringMappings = new HashMap<>();
+
+    // Construct a new List to store the Map's keys
+    List<String> keys = new ArrayList<>();
+
+    while (it.hasNext()) {
+      Map.Entry<Object, Object> current = it.next();
+
+      // Get Key & Value, and update string mappings map
+      String key = current.getKey().toString();
```

Comment:
We need to think about what to do if the map contains both the string "1" and the number 1.

Reply:
@ewpatton The default solution in place (according to LinkedHashMap order) is to take the most recent (last) entry in the map (so effectively, new values override older ones). I think this is fairly reasonable, since we are replacing values in text, and there is no distinction in text between a number literal, and a number character. It might be worth adding in the description of the block or in the MD documentation that this would be the expected behavior.

Reply:
Also might be worthwhile detecting this case explicitly, since both keys will be added to the List. Good point!